src/cdk-experimental/popover-edit/edit-event-dispatcher.ts

```diff
@@ -88,20 +182,120 @@ export class EditEventDispatcher {
+   * Hovering is defined as when the mouse has momentarily stopped moving over the cell.
    */
-  hoveringOnRow(element: Element|EventTarget): Observable<boolean> {
-    let row: Element|null = null;
-
-    return this._hoveringDistinct.pipe(
-        map(hoveredRow => hoveredRow === (row || (row = closest(element, ROW_SELECTOR)))),
-        audit(
-            (hovering) => hovering ? this.mouseMove.pipe(filter(hoveredRow => hoveredRow === row)) :
-                                     timer(DEFAULT_MOUSE_OUT_DELAY_MS)),
-        distinctUntilChanged(),
+  hoverOrFocusOnRow(row: Element): Observable<HoverContentState> {
+    if (row !== this._lastSeenRow) {
+      this._lastSeenRow = row;
+      this._lastSeenRowHoverOrFocus = this._hoveredContentStateDistinct.pipe(
+        map(state => state.get(row) || HoverContentState.OFF),
+        this._distinctShare,
+      );
+    }
+
+    return this._lastSeenRowHoverOrFocus!;
+  }
+
+  private _enterZone<T>(): MonoTypeOperatorFunction<T> {
+    return <T>(source: Observable<T>) =>
+        new Observable<T>((observer) => source.subscribe({
+             next: (value) => this._ngZone.run(() => observer.next(value)),
+             error: (err) => observer.error(err),
+             complete: () => observer.complete()
+           }));
+  }
+
+  private _getFirstRowWithHoverContent(): Observable<Element|null> {
+    return this._mapAllRowsToSingleRow(rows => {
+      for (let i = 0, row; row = rows[i]; i++) {
+        if (this._rowsWithHoverContent.has(row as Element)) {
+          return row as Element;
+        }
+      }
+      return null;
+    });
+  }
+  
+  private _getLastRowWithHoverContent(): Observable<Element|null> {
+    return this._mapAllRowsToSingleRow(rows => {
+      for (let i = rows.length - 1, row; row = rows[i]; i--) {
+        if (this._rowsWithHoverContent.has(row as Element)) {
+          return row as Element;
+        }
+      }
+      return null;
+    });
+  }
+
+  private _mapAllRowsToSingleRow(mapper: (rows: NodeList) => Element|null):
+      Observable<Element|null> {
+    return this.allRows.pipe(
+        map(mapper),
+        this._startWithNullDistinct,
     );
+  } 
+}
+
+function computeHoverContentState([firstRow, lastRow, activeRow, hoverRow]: Array<Element|null>):
+     Map<Element, HoverContentState> {
+  const hoverContentState = new Map<Element, HoverContentState>();
+
+  // Add focusable rows.
+  for (const focussableRow of [
+    firstRow,
+    lastRow,
+    activeRow && activeRow.previousElementSibling,
+    activeRow && activeRow.nextElementSibling,
+  ]) {
+    if (focussableRow) {
+      hoverContentState.set(focussableRow as Element, HoverContentState.FOCUSABLE);
+    }
   }
+
+  // Add/overwrite with fully visible rows.
+  for (const onRow of [activeRow, hoverRow]) {
+    if (onRow) {
+      hoverContentState.set(onRow, HoverContentState.ON);
+    }
+  }
+
+  return hoverContentState;
+}
+
+// TODO: use unknown rather than any.
```

Comment:
I think this could be generic instead of using `any` or `unknown`, since all it cares about is that `a` and `b` have the same key/value types
```ts
function areMapEntriesEqual<K, V>(a: Map<K, V>, b: Map<K, V>): boolean {
  if (a.size !== b.size) {
    return false;
  }

  for (const aKey of Array.from(a.keys())) {
    if (b.get(aKey) !== a.get(aKey)) {
      return false;
    }
  }
```